### PR TITLE
[FIX] website_sale_stock_provisioning_date: timezone

### PR DESCRIPTION
**Pull request by chienandalu**
We want to show the provisioning date in the context of the company
timezone. Otherwise we could get missleading dates.

cc @Tecnativa TT33244

ping @ernestotejeda @pedrobaeza

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Not sure why using such tz in a calendar that may not be assigned. I think you have to use the user timezone that is visiting your website.

**Comment by chienandalu** [comment]
Well, the problem here is the following. We get a move that has a planned date of 2022-01-01 00:00:00 (UTC+1). So the date in which the company will receive it is 2022-01-01, but as it's stored in UTC, different users could get different dates. For example, a user from Portugal (UTC+0) would see that the planned date is 2021-12-31, and that's what we want to avoid.

**Comment by pedrobaeza** [comment]
I think we are trying to solve a problem that comes from a bad user input. If a user introduces in the expected date `01/01/2022 00:00:00`, and their timezone is GMT+2, in fact the day that will come will be 31/12/2021, not the other one. It should introduce `01/01/2022 12:00:00` as the way to set exactly a realistic date (you are not going to pickup the stock on midnight). Anyways, I insist that if you want to "decipher" the date that the user is seeing when they introduce the expected date, then you must use user timezone, not a probably undefined company calendar timezone.

**Comment by chienandalu** [comment]
It will be a problem with good user inputs and large timezone differences. We want to show the date the company receives it in their timezone. We can discuss whether or not the company resource calendar is reliable or not (there aren't many alternatives, or at least not that simple)

**Comment by chienandalu** [comment]
@pedrobaeza Changes done according to what we commented internally

File: website_sale_stock_provisioning_date/hooks.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    cr.execute(
+        """
+       UPDATE stock_move sm0
+       SET date_expected_date = (
+           sm.date_expected AT TIME ZONE COALESCE(rp.tz, 'UTC')
+       )::date
+       FROM stock_move sm
+       LEFT JOIN res_users ru ON sm.write_uid = ru.id
+       LEFT JOIN res_partner rp ON ru.partner_id = rp.id
+       WHERE sm.date_expected IS NOT NULL AND sm0.id = sm.id
```
Comment (pedrobaeza) [inline website_sale_stock_provisioning_date/hooks.py]: Shouldn't you limit this to the non done moves?

File: website_sale_stock_provisioning_date/models/stock_move.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from pytz import timezone
+
+from odoo import api, fields, models
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    date_expected_date = fields.Date(
```
Comment (pedrobaeza) [inline website_sale_stock_provisioning_date/models/stock_move.py]: Isn't better to call it `date_provisioning`?

File: website_sale_stock_provisioning_date/models/stock_move.py [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+        string="Date expected", compute="_compute_date_expected_date", store=True,
+    )
+
+    @api.depends("date_expected")
+    def _compute_date_expected_date(self):
+        self.date_expected_date = False
+        tz = timezone(self.env.user.tz or "UTC")
```
Comment (pedrobaeza) [inline website_sale_stock_provisioning_date/models/stock_move.py:21]: You can put a note saying that it's expected that when the `date_expected` is changed, the user that is changing it (thus `self.env.user`), expects the day that they are seeing

File: website_sale_stock_provisioning_date/models/stock_move.py (hunk 2 above)
Comment (chienandalu) [reply website_sale_stock_provisioning_date/models/stock_move.py]: It was my initial call, but it doesn't fit when the move isn't an incoming one, I think...

Comment (pedrobaeza) [reply website_sale_stock_provisioning_date/models/stock_move.py]: Well, this field name is very weird right now. It's true the other is not the finest one, but is not bad at all I think.

File: website_sale_stock_provisioning_date/hooks.py (hunk 1 above)
Comment (chienandalu) [reply website_sale_stock_provisioning_date/hooks.py]: Sure. Done

File: website_sale_stock_provisioning_date/models/stock_move.py (hunk 3 above)
Comment (chienandalu) [reply website_sale_stock_provisioning_date/models/stock_move.py:21]: Done

**Review by pedrobaeza — APPROVED** [review]

File: website_sale_stock_provisioning_date/models/stock_move.py (hunk 2 above)
Comment (chienandalu) [reply website_sale_stock_provisioning_date/models/stock_move.py]: ok then

File: website_sale_stock_provisioning_date/hooks.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        SET date_provisioning = (
+            sm.date_expected AT TIME ZONE COALESCE(rp.tz, 'UTC')
+        )::date
+        FROM stock_move sm
+        LEFT JOIN res_users ru ON sm.write_uid = ru.id
+        LEFT JOIN res_partner rp ON ru.partner_id = rp.id
+        WHERE
+            sm0.id = sm.id
+            AND sm.date_expected IS NOT NULL
+            AND sm.state IN ('waiting', 'confirmed', 'partially_available', 'assigned')
```
Comment (pedrobaeza) [inline website_sale_stock_provisioning_date/hooks.py]: For the possible draft moves?

```suggestion
            AND sm.state != 'done
```

Comment (chienandalu) [reply website_sale_stock_provisioning_date/hooks.py]: Changed discarding done and cancelled moves

**Review by ernestotejeda — APPROVED** [review]

**Pull request merged**